content/blog/2019-12-18-dfg-cover/index.md

```diff
@@ -0,0 +1,383 @@
+we used `matmult-int.c` to chose a few common chains of operations:
+
+`mul` &rarr; `add` &rarr; `srem`
+
+`shl` &rarr; `add`
+
+`sdiv` &rarr; `mul` &rarr; `add`
+
+
+As we expected, these small human-selected stencils subgraphs performed
+especially poorly. On the original program, `matmult-int.c`, these stencils only
+matched less than 4% of instructions.
+
+[embench]: https://embench.org
+[`networkx.isomorphism`]: https://networkx.github.io/documentation/stable/reference/algorithms/isomorphism.html
+
+## Identifying common DFG stencils
+
+Of course, finding the common subgraphs by hand is pretty antithetical to a
+reasonable approach at compiling code. Our real goal is to automate the process
+of finding the common DFG stencils to accelerate.
+
+### Formal description of the task
+
+In this context of ignoring control flow and considering data flows within basic
+block, we can look at the problem purely graph-theoretically. For a single trace
+through the program, the data flow graph $G$ is acyclic, and we would like to
+cover as much of it as possible with subgraphs corresponding to the stencils
+that we accelerate. Statically, we do not know what the final data flow graph
+is, but we do know that we will be able to assemble one by connecting dangling
+edges from control-flow-free components: basic blocks.
+
+We would like to find a small collection of graph components $\mathcal H = \{H_i, \ldots, H_k\}$, which we can use to replace parts of and accelerate programs having basic blocks $\mathcal G = \{G_1,\ldots,G_n\}$, that maximizes the total saved time:
+
+$$\mathcal S_{\mathcal H}(\mathcal G) := \max_{\mathcal C \in \text{Cov}(\mathcal G, \mathcal H)}~ \sum_{G \in \mathcal G} w_G \cdot  \sum_{H \in \mathcal C_G} f_H \cdot |H|$$
+
+where:
+
+- $\text{Cov}(\mathcal G, \mathcal H)$ is the set of all valid (partial)
+    coverings of basic blocks with at most one stencil, that is, injective graph
+     morphisms $\varphi: (\cup \mathcal G) \to \cup \mathcal H$.
+- $\mathcal C_G$ is the component of the covering $\mathcal C$ of the
+    total covering on the particular basic block graph $G$.
+- $w_G$ is independent of $\mathcal H$ and proportional to the expected
+    number of times $G$ is executed.
+- $f_H$ is the expected speedup factor from accelerating the component $H$.
+
+To consider this as a general learning problem, imagine that there's some
+underlying distribution $\mathtt{Programs}$ of programs that people write; we
+can now cast our work as a solution to the optimization problem of finding:
+
+$$ \arg\max_{\mathcal H}\left( \mathop{\mathbb E}\limits_{\mathcal G\sim \texttt{Programs}}~ \mathcal S_{\mathcal H}(\mathcal G) - \text{Cost}(\mathcal H) \right)$$
+
+where $\text{Cost}(\mathcal H)$ is the additional cost incurred by choosing to
+accelerate the subgraph stencil $\mathcal H$, which is higher for larger
+subgraphs. In this learning analogy, finding common DFGs for a particular
+collection of programs is training data.
```

Comment:
Hmm, I'm not _quite_ following the analogy… is the training data DFG/program pairs? Where would the training data come from? Maybe it's just too late at night but for some reason I wasn't quite able to follow what's going on in this whole learning-based formulation…

Reply:
Sorry, I will expand this to make it clearer. The training data is a program DFG. The hypothesis class consists of possible collections $\mathcal H$ of snippets; the evaluation is the coverage of the program. You can test on your training data (see how well your snippets work for the program they were fit to) or on held-out data (generate snippets from one program, and observe their coverage on another). 

The point is that we can either stipulate the most useful snippets by hand, or learn them by example.

Reply:
Got it; that helps! Thanks!